File: Python/051__N-Queens.py
# Clean DFS Solution
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        def DFS(queens, xy_dif, xy_sum):
            p = len(queens)
            if p == n:
                result.append(queens)
                return None
            for q in range(n):
                if q not in queens and p - q not in xy_dif and p + q not in xy_sum:
                    DFS(queens + [q], xy_dif + [p - q], xy_sum + [p + q])
        result = []
        DFS([],[],[])
        return [ ["." * i + "Q" + "." * (n - i - 1) for i in sol] for sol in result]

# Backtracking Solution
class Solution:
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        if n == 0:
            return []
        self.n = n
        self.col_taken, self.diag_taken, self.antiDiag_taken = set(), set(), set()
        self.result = []
        self.search(0, [])

        return self.result

    # We will go row by row, for each row, try putting a queen on each column
    def search(self, row, path):
        if row == self.n:
            self.result.append(path)
            return

        for col in range(self.n):
            if col not in self.col_taken and col - row not in self.diag_taken and col + row not in self.antiDiag_taken:
                self.col_taken.add(col)
                self.diag_taken.add(col - row)
                self.antiDiag_taken.add(col + row)
                self.search(row + 1, path + ["." * col + "Q" + (self.n - col - 1) * "."])
                self.col_taken.remove(col)
                self.diag_taken.remove(col - row)
                self.antiDiag_taken.remove(col + row)

File: Java/051__N-Queens.java
class Solution {

    int rows[];
    int hills[];
    int dales[];
    int n;
    List<List<String>> result = new ArrayList();
    int queens[];

    public List<List<String>> solveNQueens(int n) {
        this.n = n;
        rows = new int[n];
        hills = new int[4 * n - 1];
        dales = new int[2 * n - 1];
        queens = new int[n];
        backtrack(0);
        return result;
    }

    private void backtrack(int row) {
        for (int col = 0; col < n; col++) {
            if (notUnderArrack(row, col)) {
                placeQueen(row, col);
                if (row + 1 == n)
                    addSolution();
                else
                    backtrack(row + 1);
                removeQueen(row, col);
            }
        }
    }

    private boolean notUnderArrack(int row, int col) {
        int result = rows[col] + hills[row - col + 2 * n] + dales[row + col];
        return (result == 0) ? true : false;
    }

    private void placeQueen(int row, int col) {
        queens[row] = col;
        rows[col] = 1;
        hills[row - col + 2 * n] = 1;
        dales[row + col] = 1;
    }

    private void removeQueen(int row, int col) {
        queens[row] = 0;
        rows[col] = 0;
        hills[row - col + 2 * n] = 0;
        dales[row + col] = 0;
    }

    private void addSolution() {
        List<String> solution = new ArrayList<String>();
        for (int i = 0; i < n; i++) {
          int col = queens[i];
          StringBuilder sb = new StringBuilder();
          for(int j = 0; j < col; j++)
              sb.append(".");
          sb.append("Q");
          for (int j = 0; j < n - col - 1; j++)
              sb.append(".");
          solution.add(sb.toString());
        }
        result.add(solution);
    }
}
